subxt/examples/unstable_light_client_parachains.rs

```diff
@@ -0,0 +1,101 @@
+use futures::StreamExt;
+use std::{iter, num::NonZeroU32};
+use subxt::{
+    client::{LightClient, RawLightClient},
+    PolkadotConfig,
+};
+
+// Generate an interface that we can use from the node's metadata.
+#[subxt::subxt(runtime_metadata_path = "../artifacts/polkadot_metadata_small.scale")]
+pub mod polkadot {}
+
+const POLKADOT_SPEC: &str = include_str!("../../artifacts/demo_chain_specs/polkadot.json");
+const ASSET_HUB_SPEC: &str =
+    include_str!("../../artifacts/demo_chain_specs/polkadot_asset_hub.json");
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // The smoldot logs are informative:
+    tracing_subscriber::fmt::init();
+
+    // Connecting to a parachain is a multi step process.
+
+    // Step 1. Construct a new smoldot client.
+    let mut client = subxt_lightclient::Client::new(subxt_lightclient::DefaultPlatform::new(
+        "subxt-example-light-client".into(),
+        "version-0".into(),
+    ));
+
+    // Step 2. Connect to the relay chain of the parachain. For this example, the Polkadot relay chain.
+    let polkadot_connection = client
+        .add_chain(subxt_lightclient::AddChainConfig {
+            specification: POLKADOT_SPEC,
+            json_rpc: subxt_lightclient::AddChainConfigJsonRpc::Enabled {
+                max_pending_requests: NonZeroU32::new(128).unwrap(),
+                max_subscriptions: 1024,
+            },
+            potential_relay_chains: iter::empty(),
+            database_content: "",
+            user_data: (),
+        })
+        .expect("Light client chain added with valid spec; qed");
+    let polkadot_json_rpc_responses = polkadot_connection
+        .json_rpc_responses
+        .expect("Light client configured with json rpc enabled; qed");
+    let polkadot_chain_id = polkadot_connection.chain_id;
+
+    // Step 3. Connect to the parachain. For this example, the Asset hub parachain.
+    let assethub_connection = client
+        .add_chain(subxt_lightclient::AddChainConfig {
+            specification: ASSET_HUB_SPEC,
+            json_rpc: subxt_lightclient::AddChainConfigJsonRpc::Enabled {
+                max_pending_requests: NonZeroU32::new(128).unwrap(),
+                max_subscriptions: 1024,
+            },
+            // The chain specification of the asset hub parachain mentions that the identifier
+            // of its relay chain is `polkadot`.
+            potential_relay_chains: [polkadot_chain_id].into_iter(),
+            database_content: "",
+            user_data: (),
+        })
+        .expect("Light client chain added with valid spec; qed");
+    let parachain_json_rpc_responses = assethub_connection
+        .json_rpc_responses
```

Comment:
Is it possible to write a wrapper for this API?

It's quite annoying to do this unwrap because the RPC client should always be enabled in smoldot?

Reply:
Yep, I was thinking of creating a bit of a higher level API as a followup to this.

Ideally we could provide:
- The highest level API used for most people out there, but with some API limitations (ie multichain usage). And this is already merged in subxt. Btw we need to handle this under the hood, its just the way smoldot exposes that API at the moment depending if the RPC was enabled or not.



```rust
let api = LightClient::<PolkadotConfig>::builder()
        .bootnodes([
            "/ip4/127.0.0.1/tcp/30333/p2p/12D3KooWEyoppNCUx8Yx66oV9fJnriXwCcXwDDUA2kj6vnc6iDEp",
        ])
        .build_from_url("ws://127.0.0.1:9944")
        .await?;
```

- A low level API that provides the most flexibility and allows users to utilize 100% of smoldot features. This is included in this PR, although the tradeoff is that we have some extra method calls on our builder to insert the `chain_id, rpc-responses` objects.

- I was also thinking of providing something in between were we tackle the multichain use-case together with parachain use-case. And sort of evolve the API in a natural way when smoldot adds additional features; or users may need something easier to use than the low level API. 

```rust

/// Naming tbd, suggesting that's somewhere in the middle between super high level and super low level API.
let client: MediumLightClient  = MediumLightClient::builder()
    .add_chain(
         AddChianBuilder::with_identifier( "polkadot" )
             .with_spec( include_str!( "../assets/polkadot_chain_spec.json")
    )
    .add_chain(
         AddChainBuilder::with_identifier( "asset-hub")
            .relay_chain( "polkadot")
            .from_url("ws://127.0.0.1:9944")
     )
    .build().await;
    
let polkadot_api = client.for_chain( "polkadot");
```


Something like that, but I didn't put a proper thought into the API. We still need a way to identify the chains while adding them to the client. And another approach here, would be to have a `fn add_chain(&mut self) -> SubxtChainDetails` and abstract that bit away; as opposed to allowing users to provide strs / tokens.

Another thing worth exploring would be extending the `LightClientBuilder`, maybe we could have a `LightClient::builder().from_token(PolkadotToken)` etc

Still believe having the ability for users to turn a raw light client into something that subxt can use under the scene would cover the most complex use-cases :D

Reply:
Yup, I'm generally for the idea too of having: 
- a simple high level API (like we do)
- an API that allows one to manually configure a `smoldot::Client` or whatever however they need and connect that to subxt (like this PR)
- maybe something inbetween which a `smoldot::Client` could be turned into as one way of building it, or it could be built via some nice higher level APIs that avoid the duplication of passing chain IDs around and such

But yeah it does need some thought and experimentation to find the ideal interface for this stuff I think; we shouldn't rush into it :)